### PR TITLE
Featured Artists

### PR DESCRIPTION
**Pull request by cbj4074**
#136

### CODE REVIEW

File: database/factories/ArtistFactory.php [hunk 1]
```diff
@@ -2,10 +2,12 @@
 
 use Faker\Generator as Faker;
 
-$factory->define(App\Artist::class, function (Faker $faker) {
-    return [
+use App\Contracts\ArtistRepositoryInterface;
 
-    ];
+$repo = resolve(ArtistRepositoryInterface::class);
```
Comment (poppabear8883) [inline database/factories/ArtistFactory.php]: should name this `$artist` as all dependencies will need to be resolved

File: database/factories/ArtistFactory.php [hunk 2]
```diff
@@ -2,10 +2,12 @@
...
+
+$factory->define($repo->class(), function (Faker $faker) {
+    return [];
 });
 
 $factory->state(App\Artist::class, 'onLabel', [
```
Comment (poppabear8883) [inline database/factories/ArtistFactory.php]: `$artist->class()`

File: database/factories/ArtistFactory.php [hunk 3]
```diff
@@ -14,10 +16,15 @@
     },
 ]);
 
-$factory->afterCreating(App\Artist::class, function ($artist, $faker) {
+$factory->afterCreating($repo->class(), function ($artist, $faker) use ($repo) {
     factory(App\CatalogEntity::class)->create([
```
Comment (poppabear8883) [inline database/factories/ArtistFactory.php]: `$catalogEntity->class()`

File: database/factories/ArtistFactory.php [hunk 4]
```diff
@@ -14,10 +16,15 @@
...
         'user_id' => factory(App\User::class)->create()->id,
         'catalogable_id' => $artist->id,
-        'catalogable_type' => App\Artist::class
+        'catalogable_type' => $repo->class()
+    ]);
+
+    factory(App\Profile::class)->create([
```
Comment (poppabear8883) [inline database/factories/ArtistFactory.php]: `$profile->class()`

File: app/Featured.php [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+class Featured extends Model
+{
+    public function featurable()
+    {
+        return $this->morphTo();
+    }
+
+    public function scopeArtists($query)
+    {
+        return $query->where('featurable_type', Artist::class);
```
Comment (poppabear8883) [inline app/Featured.php]: This is tightly coupled to the concrete implementation

File: database/factories/ArtistFactory.php (hunk 1 above)
Comment (cbj4074) [reply database/factories/ArtistFactory.php]: Yes indeed. Good call. Fixed in subsequent commit.

File: database/factories/ArtistFactory.php (hunk 2 above)
Comment (cbj4074) [reply database/factories/ArtistFactory.php]: Fixed,

File: database/factories/ArtistFactory.php (hunk 3 above)
Comment (cbj4074) [reply database/factories/ArtistFactory.php]: Fixed.

File: database/factories/ArtistFactory.php (hunk 4 above)
Comment (cbj4074) [reply database/factories/ArtistFactory.php]: Aaand, fixed.

File: app/Artist.php [hunk 6]
```diff
@@ -37,4 +39,24 @@ public function user()
...
+        return $query->has('profile')
+            ->whereHas('albums', function ($query) {
+                $query->where('is_active', 1);
+            })
+            ->whereDoesntHave('featureds', function ($query) {
+                // "... where created_at is more recent than 180 days ago."
+                // The goal is to exclude Artists who have been Featured in the
+                // past 180 days.
+
+                $query->where('created_at', '>', Carbon::now()->subDays(180)->toDateTimeString());
```
Comment (poppabear8883) [inline app/Artist.php]: Does this `$query` the featured table ?

Comment (cbj4074) [reply app/Artist.php]: Yes.

File: app/Artist.php [hunk 7]
```diff
@@ -37,4 +43,24 @@ public function user()
     {
         return $this->catalogable->user();
     }
+
+    public function scopeFeaturable(Builder $query): Builder
+    {
+        return $query->has('profile')
```
Comment (poppabear8883) [inline app/Artist.php]: This query will always be `true` due to the fact that when an artist is created via the `create()` method a profile is also created as seen here https://github.com/indiehd/web-api/blob/master/app/Repositories/ArtistRepository.php#L44

Maybe its best to remove the profile rule all together, I was thinking initially that the CatelogEntity had nullable fields that we would require to have values in order to be featured. However i am now seeing that i was mistaken (or confused). Lets just eliminate this particular query unless you feel that there is a good reason (or alternative proposal) to keep this in ?

File: tests/Feature/Controllers/FeaturedControllerTest.php [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+        $this->album = resolve(AlbumRepositoryInterface::class);
+    }
+
+    public function testAllReturnsOkStatusAndExpectedJsonStructure()
+    {
+        $artist = $this->createArtist();
+
+        $this->makeAlbum([
+            'artist_id' => $artist->id,
+            'is_active' => true,
```
Comment (poppabear8883) [inline tests/Feature/Controllers/FeaturedControllerTest.php]: Didn't you remove the `is_active` from the structure ?

File: app/Artist.php (hunk 7 above)
Comment (cbj4074) [reply app/Artist.php]: An excellent point. I removed it and all tests still pass, which says quite enough. 👍

File: tests/Feature/Controllers/FeaturedControllerTest.php (hunk 8 above)
Comment (cbj4074) [reply tests/Feature/Controllers/FeaturedControllerTest.php]: I remove it from the `Featured` model structure, not from the `Album`, which is what we're looking at here.

**Pull request merged**